tagstudio/src/qt/modals/drop_import.py

```diff
@@ -0,0 +1,251 @@
+from pathlib import Path
+import shutil
+import typing
+
+from PySide6.QtCore import QThreadPool, Qt, QMimeData, QUrl
+from PySide6.QtGui import (
+    QDropEvent,
+    QDragEnterEvent,
+    QDragMoveEvent,
+    QMouseEvent,
+    QDrag,
+    QDragLeaveEvent,
+)
+from PySide6.QtWidgets import QMessageBox
+from src.qt.widgets.progress import ProgressWidget
+from src.qt.helpers.custom_runnable import CustomRunnable
+from src.qt.helpers.function_iterator import FunctionIterator
+
+if typing.TYPE_CHECKING:
+    from src.qt.ts_qt import QtDriver
+
+import logging
+
+
+class DropImport:
+    def __init__(self, driver: "QtDriver"):
+        self.driver = driver
+
+    def mouseMoveEvent(self, event: QMouseEvent):
+        if event.buttons() is not Qt.MouseButton.LeftButton:
+            return
+        if len(self.driver.selected) == 0:
+            return
+
+        drag = QDrag(self.driver)
+        paths = []
+        mimedata = QMimeData()
+        for selected in self.driver.selected:
+            entry = self.driver.lib.get_entry(selected[1])
+            url = QUrl.fromLocalFile(
+                Path(self.driver.lib.library_dir) / entry.path / entry.filename
+            )
+            paths.append(url)
+
+        mimedata.setUrls(paths)
+        drag.setMimeData(mimedata)
+        drag.exec(Qt.DropAction.CopyAction)
+
+    def dropEvent(self, event: QDropEvent):
+        if (
+            event.source() is self.driver
+        ):  # change that if you want to drop something originating from tagstudio, for moving or so
+            return
+
+        if not event.mimeData().hasUrls():
+            return
+
+        self.urls = event.mimeData().urls()
+        self.import_files()
+
+    def dragLeaveEvent(self, event: QDragLeaveEvent):
+        if event.mimeData().hasUrls():
+            event.accept()
+        else:
+            event.ignore()
+
+    def dragEnterEvent(self, event: QDragEnterEvent):
+        if event.mimeData().hasUrls():
+            event.accept()
+        else:
+            event.ignore()
+
+    def dragMoveEvent(self, event: QDragMoveEvent):
+        if event.mimeData().hasUrls():
+            event.accept()
+        else:
+            logging.info(self.driver.selected)
+            event.ignore()
+
+    def import_files(self):
+        self.files: list[Path] = []
+        self.dirs_in_root: list[Path] = []
+        self.duplicate_files: list[Path] = []
+
+        create_progress_bar(
+            self.collect_files_to_import,
+            "Searching Files",
+            "Searching New Files...\nPreparing...",
+            lambda x: f'Searching New Files...\n{x[0]+1} File{'s' if x[0]+1 != 1 else ''} Found. {(f'{x[1]} Already exist in the library folders') if x[1]>0 else ''}',
```

Comment:
it should be enough to use double quotes on the "edges":
```suggestion
            lambda x: f"Searching New Files...\n{x[0]+1} File{'s' if x[0]+1 != 1 else ''} Found. {(f'{x[1]} Already exist in the library folders') if x[1]>0 else ''}",
```

Generally I'd recommend to extract this into a standalone function because this is way too much logic in one f-string.